File: Contact_PSTH/psth_raster_whisker_contact.py
#!/usr/bin/env python3
"""
PSTH + raster aligned to whisker contact interval onsets, per unit.

Spikes CSV: column 0 = spike time (seconds), column 1 = unit (cluster id).
Intervals CSV: Start, End = frame numbers (kept as-is).
Time alignment: spike time (s) is multiplied by FPS (500) to get spike time in frame
units; relative time for each interval is (spike_frames - Start) / FPS in seconds.

Usage: python psth_raster_whisker_contact.py [interval_csv] [spikes_csv]
"""

import argparse
import csv
from pathlib import Path

import numpy as np
import matplotlib.pyplot as plt

FPS = 500.0
BIN_S = 0.01   # 10 ms bins
T_PRE = 0.3    # seconds before interval start
T_POST = 2   # seconds after interval start


def load_intervals(path):
    """Load interval CSV. Start/End are frame numbers (kept as-is, no division)."""
    rows = []
    with open(path, newline="") as f:
        r = csv.reader(f)
        header = next(r)
        for row in r:
            if not row:
                continue
            start_fr = int(row[0].strip())
            end_fr = int(row[1].strip())
            rows.append({"start_fr": start_fr, "end_fr": end_fr})
    return rows


def load_spikes(path):
    """
    Load spikes CSV.
    Column 0 = spike time in seconds.
    Column 1 = unit (cluster id).
    Column 2 = ignored.
    """
    times = []
    units = []
    with open(path, newline="") as f:
        r = csv.reader(f, skipinitialspace=True)
        for row in r:
            if len(row) < 2:
                continue
            t = float(row[0].strip())
            unit = int(row[1].strip())
            times.append(t)
            units.append(unit)
    return np.array(times), np.array(units)


def psth_for_unit(spike_frames, spike_units, intervals, unit, bin_s=BIN_S, t_pre=T_PRE, t_post=T_POST):
    """PSTH aligned to each interval start. spike_frames = spike time (s) * FPS. Returns bin_centers, rate, n_trials."""
    mask = spike_units == unit
    unit_frames = spike_frames[mask]

    t_edges = np.arange(-t_pre, t_post + 1e-9, bin_s)
    n_bins = len(t_edges) - 1
    bin_centers = (t_edges[:-1] + t_edges[1:]) / 2
    n_trials = len(intervals)
    counts = np.zeros(n_bins)

    for iv in intervals:
        t0_fr = iv["start_fr"]
        for sf in unit_frames:
            rel_fr = sf - t0_fr
            rel_s = rel_fr / FPS
            if -t_pre <= rel_s < t_post:
                idx = np.searchsorted(t_edges, rel_s, side="right") - 1
                idx = np.clip(idx, 0, n_bins - 1)
                counts[idx] += 1

    rate = counts / (n_trials * bin_s)
    return bin_centers, rate, n_trials


def raster_for_unit(spike_frames, spike_units, intervals, unit, t_pre=T_PRE, t_post=T_POST):
    """Relative spike times (s) and trial index. spike_frames = spike time (s) * FPS."""
    mask = spike_units == unit
    unit_frames = spike_frames[mask]

    rel_times = []
    trial_inds = []

    for i, iv in enumerate(intervals):
        t0_fr = iv["start_fr"]
        for sf in unit_frames:
            rel_fr = sf - t0_fr
            rel_s = rel_fr / FPS
            if -t_pre <= rel_s < t_post:
                rel_times.append(rel_s)
                trial_inds.append(i + 1)  # 1-based for plot

    return np.array(rel_times), np.array(trial_inds)


def main():
    base = Path("/Users/martindokholyan/Desktop/Plots/2")
    parser = argparse.ArgumentParser(description="PSTH + raster per unit, aligned to whisker contact onset.")
    parser.add_argument("interval_csv", nargs="?", default=str(base / "interval_2_whisker_contact.csv"),
                        help="CSV with Start,End frame numbers")
    parser.add_argument("spikes_csv", nargs="?", default=str(base / "spikes.csv"),
                        help="CSV: col0=time(s), col1=unit")
    args = parser.parse_args()

    interval_path = Path(args.interval_csv)
    spikes_path = Path(args.spikes_csv)
    out_dir = interval_path.parent

    # Intervals: keep Start/End as frame numbers (no division)
    intervals = load_intervals(interval_path)
    # For raster display, sort intervals by duration (shortest to longest)
    intervals_sorted = sorted(
        intervals,
        key=lambda iv: iv["end_fr"] - iv["start_fr"],
    )
    # Spikes: col0 = time in seconds, col1 = unit. Convert to frame units: multiply by FPS
    spike_times_s, spike_units = load_spikes(spikes_path)
    spike_frames = spike_times_s * FPS
    units = np.unique(spike_units)
    units = np.sort(units)

    for unit in units:
        # PSTH is invariant to trial order, so we can keep original interval order here
        bin_centers, rate, n_trials = psth_for_unit(spike_frames, spike_units, intervals, unit)
        # Raster rows ordered by interval duration via intervals_sorted
        rel_times, trial_inds = raster_for_unit(spike_frames, spike_units, intervals_sorted, unit)

        fig, (ax_psth, ax_raster) = plt.subplots(2, 1, sharex=True, figsize=(6.4, 5.6),
                                                 gridspec_kw=dict(height_ratios=[1, 1.2]))

        # PSTH
        ax_psth.plot(bin_centers, rate, color="steelblue", lw=2)
        ax_psth.axvline(0, color="black", ls="--")
        ax_psth.set_ylabel("Rate (spikes/s)")
        ax_psth.set_title(f"PSTH — Unit {unit} (n={n_trials} intervals)")
        ax_psth.set_xlim(-T_PRE, T_POST)

        # Raster: draw a short vertical tick within each row (trial)
        if len(rel_times) > 0:
            tick_half = 0.4  # half-height of tick within each trial row
            x_seg = []
            y_seg = []
            for t, tr in zip(rel_times, trial_inds):
                x_seg.extend([t, t, np.nan])
                y_seg.extend([tr - tick_half, tr + tick_half, np.nan])
            ax_raster.plot(x_seg, y_seg, color="black", linewidth=0.8)
        ax_raster.axvline(0, color="black", ls="--")
        ax_raster.set_xlabel("Time from contact onset (s)")
        ax_raster.set_ylabel("Trial")
        ax_raster.set_title("Raster")
        ax_raster.set_ylim(0.5, n_trials + 0.5)
        ax_raster.set_xlim(-T_PRE, T_POST)

        plt.tight_layout()
        out_file = out_dir / f"psth_raster_unit_{unit}.png"
        plt.savefig(out_file, dpi=120)
        plt.close()
        print(f"Saved {out_file}")


if __name__ == "__main__":
    main()

File: Contact_PSTH/psth_raster_whisker_contact.jl
#!/usr/bin/env julia
# PSTH + raster aligned to whisker contact interval onsets, per unit.
# Usage: julia psth_raster_whisker_contact.jl [interval_csv] [spikes_csv]

using CSV
using DataFrames
using Plots
using Statistics

const FPS = 350.0
const BIN_S = 0.01   # 10 ms bins
const T_PRE = 0.1    # seconds before interval start
const T_POST = 0.3   # seconds after interval start

function load_intervals(path)
    df = CSV.read(path, DataFrame; types=Dict("Start" => Int, "End" => Int))
    df[!, :start_s] = df.Start ./ FPS
    df[!, :end_s] = df.End ./ FPS
    df
end

function load_spikes(path)
    df = CSV.read(path, DataFrame; header=["time", "unit", "ignore"],
                 types=[Float64, Int, Int], delim=',', ignorerepeated=true)
    df
end

function psth_for_unit(spikes, intervals, unit; bin_s=BIN_S, t_pre=T_PRE, t_post=T_POST)
    t_edges = (-t_pre):bin_s:(t_post + 1e-9)
    n_bins = length(t_edges) - 1
    bin_centers = (t_edges[1:end-1] .+ t_edges[2:end]) ./ 2
    unit_spikes = spikes[spikes.unit .== unit, :].time
    n_trials = nrow(intervals)
    counts = zeros(n_bins)

    for row in eachrow(intervals)
        t0 = row.start_s
        for t in unit_spikes
            rel = t - t0
            if -t_pre <= rel < t_post
                idx = searchsortedlast(t_edges, rel)
                idx = clamp(idx, 1, n_bins)
                counts[idx] += 1
            end
        end
    end

    rate = counts ./ (n_trials * bin_s)
    (; bin_centers, rate, counts, n_trials)
end

"""Return (trial_indices, relative_times) for raster: one spike per entry."""
function raster_for_unit(spikes, intervals, unit; t_pre=T_PRE, t_post=T_POST)
    unit_spikes = spikes[spikes.unit .== unit, :].time
    rel_times = Float64[]
    trial_inds = Int[]

    for (i, row) in enumerate(eachrow(intervals))
        t0 = row.start_s
        for t in unit_spikes
            rel = t - t0
            if -t_pre <= rel < t_post
                push!(rel_times, rel)
                push!(trial_inds, i)
            end
        end
    end

    (; trial_inds, rel_times)
end

function main()
    base = "/Users/martindokholyan/Desktop"
    interval_path = length(ARGS) >= 1 ? ARGS[1] : joinpath(base, "interval_0_whisker_contact.csv")
    spikes_path   = length(ARGS) >= 2 ? ARGS[2] : joinpath(base, "spikes.csv")

    intervals = load_intervals(interval_path)
    spikes = load_spikes(spikes_path)
    units = sort(unique(spikes.unit))

    for unit in units
        p = psth_for_unit(spikes, intervals, unit)
        r = raster_for_unit(spikes, intervals, unit)

        plt = plot(layout=(2, 1), size=(600, 500), link=:x,
                   xlims=(-T_PRE, T_POST))

        # Top: PSTH
        plot!(plt[1], p.bin_centers, p.rate; label="Unit $unit",
              xlabel="", ylabel="Rate (spikes/s)",
              title="PSTH — Unit $unit (n=$(p.n_trials) intervals)")
        vline!(plt[1], [0.0]; label="", color=:black, ls=:dash)

        # Bottom: Raster (vertical tick per spike)
        if !isempty(r.rel_times)
            tick_half = 0.4
            n = length(r.rel_times)
            x_seg = Vector{Float64}(undef, 3n)
            y_seg = Vector{Float64}(undef, 3n)
            for (i, (t, tr)) in enumerate(zip(r.rel_times, r.trial_inds))
                j = 3(i - 1) + 1
                x_seg[j] = x_seg[j+1] = t; x_seg[j+2] = NaN
                y_seg[j] = tr - tick_half; y_seg[j+1] = tr + tick_half; y_seg[j+2] = NaN
            end
            plot!(plt[2], x_seg, y_seg; linecolor=:black, linewidth=1, label="")
        end
        plot!(plt[2]; xlabel="Time from contact onset (s)", ylabel="Trial",
              title="Raster", ylims=(0.5, p.n_trials + 0.5))
        vline!(plt[2], [0.0]; label="", color=:black, ls=:dash)

        out_dir = dirname(interval_path)
        savefig(plt, joinpath(out_dir, "psth_raster_unit_$(unit).png"))
        println("Saved psth_raster_unit_$(unit).png")
    end
end

main()
